tools/rpctypegen/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+extern crate proc_macro;
+extern crate proc_macro2;
+
+use proc_macro::TokenStream;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+use syn::{
+    parse_macro_input, Data, DataEnum, DataStruct, DeriveInput, Fields, FieldsNamed, FieldsUnnamed,
+    Lit, Meta, MetaNameValue,
+};
+
+#[cfg(feature = "dump_errors_schema")]
+fn merge(a: &mut Value, b: &Value) {
+    match (a, b) {
+        (&mut Value::Object(ref mut a), &Value::Object(ref b)) => {
+            for (k, v) in b {
+                merge(a.entry(k.clone()).or_insert(Value::Null), v);
+            }
+        }
+        (a, b) => {
+            *a = b.clone();
+        }
+    }
+}
+
+#[derive(Default, Debug, Deserialize, Serialize)]
+struct Schema {
+    pub schema: BTreeMap<String, ErrorType>,
+}
+
+#[cfg(feature = "dump_errors_schema")]
+impl Drop for Schema {
+    fn drop(&mut self) {
+        // std::env::var("CARGO_TARGET_DIR") doesn't exists
+        let filename = "./target/rpc_errors_schema.json";
+        let schema_json = serde_json::to_value(self).expect("Schema serialize failed");
+        let new_schema_json = if let Ok(data) = std::fs::read(filename) {
+            // merge to the existing file
+            let mut existing_schema = serde_json::from_slice::<Value>(&data)
+                .expect("cannot deserialize target/existing_schema.json");
+            merge(&mut existing_schema, &schema_json);
+            existing_schema
+        } else {
+            schema_json
+        };
+        let new_schema_json_string = serde_json::to_string_pretty(&new_schema_json)
+            .expect("error schema serialization failed");
+        std::fs::write(filename, new_schema_json_string)
+            .expect("Unable to save the errors schema file");
+    }
+}
+
+#[derive(Default, Debug, Deserialize, Serialize)]
+struct ErrorType {
+    /// A type name of the error
+    pub name: String,
+    /// Names of subtypes of the error
+    pub subtypes: Vec<String>,
+    /// An error input name and a type
+    pub props: BTreeMap<String, String>,
+}
+
+thread_local!(static SCHEMA: RefCell<Schema> = RefCell::new(Schema::default()));
+
+fn parse_rpc_error_variant(input: &DeriveInput) -> Option<String> {
+    input.attrs.iter().find_map(|attr| {
+        if !attr.path.is_ident("rpc_error_variant") {
+            return None;
+        }
+        match attr.parse_meta().unwrap() {
+            Meta::NameValue(MetaNameValue { lit: Lit::Str(lit), .. }) => Some(lit.value()),
+            _ => None,
+        }
+    })
+}
+
+fn error_type_name<'a>(
+    schema: &'a mut BTreeMap<String, ErrorType>,
+    name: String,
+) -> &'a mut ErrorType {
+    let error_type = ErrorType { name: name.clone(), ..Default::default() };
+    schema.entry(name.clone()).or_insert(error_type)
+}
+
+fn parse_error_type(schema: &mut BTreeMap<String, ErrorType>, input: &DeriveInput) {
```

Comment:
Similarly to how it is done with `near-bindgen` and `borsh` you need to extract everything  but `pub fn rpc_error(input: TokenStream) -> TokenStream` into a separate crate, e.g. `rpctypegen-core` which is going to be a normal library which will be using `proc_macro2` which will allow us to have unit tests for this stuff like these: https://github.com/nearprotocol/near-bindgen/blob/master/near-bindgen-core/src/callback_args.rs#L56 Please add tests.